### PR TITLE
systemd-growfs: use sd_device_new_from_devnum()

### PR DESCRIPTION
**Pull request by tewarid**
Use sd_device_new_from_devnum() instead of device_path_make_major_minor_sysfs().

See #21592 and #24418.

### CODE REVIEW

File: src/partition/growfs.c [hunk 1]
```diff
@@ -25,13 +24,15 @@
 #include "parse-util.h"
 #include "pretty-print.h"
 #include "resize-fs.h"
+#include "sd-device.h"
```
Comment (yuwata) [inline src/partition/growfs.c]: Please move `sd-device.h` earlier place. See our [coding style](https://github.com/systemd/systemd/blob/e109541fe6988fd0e73b92226f36f41abc829883/docs/CODING_STYLE.md?plain=1#L124-L135) and other files which include sd-device.h:
```
#include <...>
#include <...>

#include "sd-device.h"

#include "blockdev-util.h"
#include "btrfs-util.h"
...
```

File: src/partition/growfs.c [hunk 2]
```diff
@@ -25,13 +24,15 @@
...
 
 static const char *arg_target = NULL;
 static bool arg_dry_run = false;
 
 #if HAVE_LIBCRYPTSETUP
 static int resize_crypt_luks_device(dev_t devno, const char *fstype, dev_t main_devno) {
-        _cleanup_free_ char *devpath = NULL, *main_devpath = NULL;
+        _cleanup_(sd_device_unrefp) sd_device *main_dev = NULL, *dev = NULL;
+        _cleanup_free_ const char *devpath = NULL, *main_devpath = NULL;
```
Comment (yuwata) [inline src/partition/growfs.c]: Please do not use `_cleanup_free_` attributes for the paths obtained by `sd_device_get_devpath()` are owned by the sd_device object.

File: src/partition/growfs.c [hunk 3]
```diff
@@ -41,22 +42,32 @@ static int resize_crypt_luks_device(dev_t devno, const char *fstype, dev_t main_
         if (r < 0)
                 return log_error_errno(r, "Cannot resize LUKS device: %m");
 
-        r = device_path_make_major_minor_sysfs(S_IFBLK, main_devno, &main_devpath);
+        r = sd_device_new_from_devnum(&main_dev, 'b', main_devno);
         if (r < 0)
-                return log_error_errno(r, "Failed to format device major/minor path: %m");
+                return log_error_errno(r, "Failed to create main sd-device for block device %u:%u: %m",
+                                       major(main_devno), minor(main_devno));
```
Comment (yuwata) [inline src/partition/growfs.c]: Please use `DEVNUM_FORMAT_STR` and `DEVNUM_FORMAT_VAL()` macros.

File: src/partition/growfs.c [hunk 4]
```diff
@@ -41,22 +42,32 @@ static int resize_crypt_luks_device(dev_t devno, const char *fstype, dev_t main_
...
 
-        main_devfd = open(main_devpath, O_RDONLY|O_CLOEXEC);
+        r = sd_device_get_devpath(main_dev, &main_devpath);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get main devpath: %m");
```
Comment (yuwata) [inline src/partition/growfs.c]: Please use `log_device_error_errno(main_dev, r, "Failed to ...: %m");`

File: src/partition/growfs.c [hunk 5]
```diff
@@ -41,22 +42,32 @@ static int resize_crypt_luks_device(dev_t devno, const char *fstype, dev_t main_
...
+
+        main_devfd = sd_device_open(main_dev, O_RDONLY|O_CLOEXEC);
         if (main_devfd < 0)
-                return log_error_errno(errno, "Failed to open \"%s\": %m", main_devpath);
+                return log_error_errno(main_devfd, "Failed to open block device \"%s\": %m", main_devpath);
```
Comment (yuwata) [inline src/partition/growfs.c]: Also here. Please use `log_device_error_errno()`.

File: src/partition/growfs.c [hunk 6]
```diff
@@ -41,22 +42,32 @@ static int resize_crypt_luks_device(dev_t devno, const char *fstype, dev_t main_
...
         if (ioctl(main_devfd, BLKGETSIZE64, &size) != 0)
                 return log_error_errno(errno, "Failed to query size of \"%s\" (before resize): %m",
                                        main_devpath);
 
         log_debug("%s is %"PRIu64" bytes", main_devpath, size);
-        r = device_path_make_major_minor_sysfs(S_IFBLK, devno, &devpath);
+        r = sd_device_new_from_devnum(&dev, 'b', devno);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create sd-device for block device %u:%u: %m",
+                                       major(devno), minor(devno));
```
Comment (yuwata) [inline src/partition/growfs.c]: Please use the macros for formatting devnum.

File: src/partition/growfs.c [hunk 7]
```diff
@@ -90,7 +101,9 @@ static int maybe_resize_underlying_device(
                 const char *mountpath,
                 dev_t main_devno) {
 
-        _cleanup_free_ char *fstype = NULL, *devpath = NULL;
+        _cleanup_(sd_device_unrefp) sd_device *dev = NULL;
+        _cleanup_free_ char *fstype = NULL;
+        _cleanup_free_ const char *devpath = NULL;
```
Comment (yuwata) [inline src/partition/growfs.c]: Also here. Please do not free `devpath`.

**Review by yuwata — CHANGES_REQUESTED** [review]
Thank you! Basically good. But several comments.

Not a big matter, but please append `()` for function names in the commit message.

File: src/partition/growfs.c [hunk 8]
```diff
@@ -109,15 +125,20 @@ static int maybe_resize_underlying_device(
                 return log_error_errno(SYNTHETIC_ERRNO(ENODEV), "File system \"%s\" not backed by block device.", arg_target);
 
         log_debug("Underlying device %d:%d, main dev %d:%d, %s",
```
Comment (yuwata) [inline src/partition/growfs.c]: Please also use `DEVNUM_FORMAT_STR` macro here.

File: src/partition/growfs.c [hunk 9]
```diff
@@ -41,22 +44,33 @@ static int resize_crypt_luks_device(dev_t devno, const char *fstype, dev_t main_
...
-        r = device_path_make_major_minor_sysfs(S_IFBLK, devno, &devpath);
+        r = sd_device_new_from_devnum(&dev, 'b', devno);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create sd-device for block device " DEVNUM_FORMAT_STR ": %m",
+                                       DEVNUM_FORMAT_VAL(devno));
+
+        r = sd_device_get_devpath(dev, &devpath);
         if (r < 0)
-                return log_error_errno(r, "Failed to format major/minor path: %m");
+                return log_error_errno(r, "Failed to get devpath: %m");
```
Comment (yuwata) [inline src/partition/growfs.c]: `log_device_error_errno()`

File: src/partition/growfs.c [hunk 10]
```diff
@@ -109,15 +125,20 @@ static int maybe_resize_underlying_device(
...
-        r = device_path_make_major_minor_sysfs(S_IFBLK, devno, &devpath);
+        r = sd_device_new_from_devnum(&dev, 'b', devno);
         if (r < 0)
-                return log_error_errno(r, "Failed to format device major/minor path: %m");
+                return log_error_errno(r, "Failed to create sd-device for block device " DEVNUM_FORMAT_STR ": %m",
+                                       DEVNUM_FORMAT_VAL(devno));
+
+        r = sd_device_get_devpath(dev, &devpath);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get devpath: %m");
```
Comment (yuwata) [inline src/partition/growfs.c]: `log_device_error_errno()`

File: src/partition/growfs.c [hunk 11]
```diff
@@ -237,13 +259,18 @@ static int run(int argc, char *argv[]) {
...
-        r = device_path_make_major_minor_sysfs(S_IFBLK, devno, &devpath);
+        r = sd_device_new_from_devnum(&dev, 'b', devno);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create sd-device for block device " DEVNUM_FORMAT_STR ": %m",
+                                       DEVNUM_FORMAT_VAL(devno));
+
+        r = sd_device_get_devpath(dev, &devpath);
         if (r < 0)
-                return log_error_errno(r, "Failed to format device major/minor path: %m");
+                return log_error_errno(r, "Failed to get devpath: %m");
```
Comment (yuwata) [inline src/partition/growfs.c]: `log_device_error_errno()`

File: src/partition/growfs.c [hunk 12]
```diff
@@ -31,7 +33,8 @@ static bool arg_dry_run = false;
 
 #if HAVE_LIBCRYPTSETUP
 static int resize_crypt_luks_device(dev_t devno, const char *fstype, dev_t main_devno) {
-        _cleanup_free_ char *devpath = NULL, *main_devpath = NULL;
+        _cleanup_(sd_device_unrefp) sd_device *main_dev = NULL, *dev = NULL;
+        const char *devpath = NULL, *main_devpath = NULL;
```
Comment (yuwata) [inline src/partition/growfs.c]: It is not necessary to initialize `devpath` and `main_devpath` with `NULL`. Please drop them.

File: src/partition/growfs.c [hunk 13]
```diff
@@ -90,7 +104,9 @@ static int maybe_resize_underlying_device(
                 const char *mountpath,
                 dev_t main_devno) {
 
-        _cleanup_free_ char *fstype = NULL, *devpath = NULL;
+        _cleanup_(sd_device_unrefp) sd_device *dev = NULL;
+        _cleanup_free_ char *fstype = NULL;
+        const char *devpath = NULL;
```
Comment (yuwata) [inline src/partition/growfs.c]: Also here. Please drop the initialization for `devpath`.

File: src/partition/growfs.c [hunk 14]
```diff
@@ -202,8 +223,9 @@ static int parse_argv(int argc, char *argv[]) {
 }
 
 static int run(int argc, char *argv[]) {
+        _cleanup_(sd_device_unrefp) sd_device *dev = NULL;
         _cleanup_close_ int mountfd = -1, devfd = -1;
-        _cleanup_free_ char *devpath = NULL;
+        const char *devpath = NULL;
```
Comment (yuwata) [inline src/partition/growfs.c]: Also here. Unnecessary initialization for `devpath`.

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks much better. But one more round please.

**Comment by poettering** [comment]
excellent. thank you!

**Review by yuwata — APPROVED** [review]
LGTM. Thank you! (and sorry for late review for the previous PR)

**Pull request merged**